crates/uv/tests/pip_compile.rs

```diff
@@ -5897,3 +5897,163 @@ fn unnamed_https_requirement() -> Result<()> {
+    let requirements_in = context.temp_dir.child("requirements.in");
+    requirements_in.write_str("anyio ; sys_platform == 'linux'")?;
+
+    uv_snapshot!(context
+        .compile()
+        .arg("requirements.in")
+        .arg("--emit-marker-expression"), @r###"
+    success: true
+    exit_code: 0
+    ----- stdout -----
+    # This file was autogenerated by uv via the following command:
+    #    uv pip compile --cache-dir [CACHE_DIR] --exclude-newer 2024-03-25T00:00:00Z requirements.in --emit-marker-expression
+    # Pinned dependencies known to be valid for:
+    #    python_version == '3.12' and platform_python_implementation == 'CPython' and sys_platform == 'linux' and platform_system == 'Linux'
+    anyio==4.3.0
+    idna==3.6
+        # via anyio
+    sniffio==1.3.1
+        # via anyio
+
+    ----- stderr -----
+    Resolved 3 packages in [TIME]
+    "###);
+
+    Ok(())
+}
+
+/// This tests that the marker expression emitted accounts for markers directly
+/// in `requirements.in`, even when the marker evaluates to false on the
+/// current platform. In this case, we set `sys_platform == 'macos'` so that on
+/// Linux, this dependency is ignored. But the marker expression generated must
+/// have `sys_platform == 'Linux'`, since the locked set of packages might be
+/// different (and indeed are different) on other platforms.
+///
+/// NOTE: This test runs on `linux` because it requires that `sys_platform
+/// == 'macos'` evaluates to `false`. While this technically only requires
+/// `not(target_os = "macos")`, the marker expression generated during test
+/// development was on Linux. So we require Linux.
+#[cfg(target_os = "linux")]
+#[test]
+fn emit_marker_expression_conditional() -> Result<()> {
+    let context = TestContext::new("3.12");
+    let requirements_in = context.temp_dir.child("requirements.in");
+    requirements_in.write_str("anyio ; sys_platform == 'macos'")?;
+
+    uv_snapshot!(context
+        .compile()
+        .arg("requirements.in")
+        .arg("--emit-marker-expression"), @r###"
+    success: true
+    exit_code: 0
+    ----- stdout -----
+    # This file was autogenerated by uv via the following command:
+    #    uv pip compile --cache-dir [CACHE_DIR] --exclude-newer 2024-03-25T00:00:00Z requirements.in --emit-marker-expression
+    # Pinned dependencies known to be valid for:
+    #    sys_platform == 'linux'
+
+    ----- stderr -----
+    Resolved 0 packages in [TIME]
+    "###);
+
+    Ok(())
+}
+
+/// This tests the marker expressions emitted when depending on a package a
+/// non-pypy dependency. Specifically, `pendulum` depends on `time-machine`,
+/// but not when using pypy.
+///
+/// NOTE: This test runs on `linux` because it was written on Linux. While the
+/// marker expression itself doesn't have anything in it that couples it to
+/// Linux, it is possible for the resolution to change on other platforms. For
+/// example, on Windows, the `tzdata` dependency is excluded. (It's actually
```

Comment:
Do you have reproduction instructions for this? That looks like a bug (https://inspector.pypi.io/project/pendulum/3.0.0/packages/67/5e/e646afbd1632bfbacdae79289d7d5879efdeeb5f5e58327bc5c698731107/pendulum-3.0.0-cp310-none-win_amd64.whl/pendulum-3.0.0.dist-info/METADATA#line.12)

Reply:
ref https://github.com/astral-sh/uv/pull/2693